nomad/worker.go

```diff
@@ -70,70 +113,286 @@ type Worker struct {
 	snapshotIndex uint64
 }
 
-// NewWorker starts a new worker associated with the given server
-func NewWorker(srv *Server) (*Worker, error) {
+// NewWorker starts a new scheduler worker associated with the given server
+func NewWorker(ctx context.Context, srv *Server, args SchedulerWorkerPoolArgs) (*Worker, error) {
+	w, _ := newWorker(ctx, srv, args)
+	w.Start()
+	return w, nil
+}
+
+// _newWorker creates a worker without calling its Start func. This is useful for testing.
+func newWorker(ctx context.Context, srv *Server, args SchedulerWorkerPoolArgs) (*Worker, error) {
 	w := &Worker{
-		srv:    srv,
-		logger: srv.logger.ResetNamed("worker"),
-		start:  time.Now(),
+		id:                uuid.Generate(),
+		srv:               srv,
+		start:             time.Now(),
+		status:            WorkerStarting,
+		enabledSchedulers: make([]string, len(args.EnabledSchedulers)),
 	}
+	copy(w.enabledSchedulers, args.EnabledSchedulers)
+
+	w.logger = srv.logger.ResetNamed("worker").With("worker_id", w.id)
 	w.pauseCond = sync.NewCond(&w.pauseLock)
-	go w.run()
+	w.ctx, w.cancelFn = context.WithCancel(ctx)
+
 	return w, nil
 }
 
-// SetPause is used to pause or unpause a worker
-func (w *Worker) SetPause(p bool) {
-	w.pauseLock.Lock()
-	w.paused = p
-	w.pauseLock.Unlock()
-	if !p {
+// ID returns a string ID for the worker.
+func (w *Worker) ID() string {
+	return w.id
+}
+
+// Start transitions a worker to the starting state. Check
+// to see if it paused using IsStarted()
+func (w *Worker) Start() {
+	w.setStatus(WorkerStarting)
+	go w.run()
+}
+
+// Pause transitions a worker to the pausing state. Check
+// to see if it paused using IsPaused()
+func (w *Worker) Pause() {
+	w.setStatus(WorkerPausing)
+	w.setPauseFlag(true)
+}
+
+// Resume transitions a worker to the resuming state. Check
+// to see if the worker restarted by calling IsStarted()
+func (w *Worker) Resume() {
+	if w.IsPaused() {
+		w.setStatus(WorkerResuming)
+		w.setPauseFlag(false)
```

Comment:
I really like the API you have here where you're setting the state as "starting/pausing/resuming" rather than "started/paused/resumed", as it draws attention to the fact that pausing is an asynchronous operation for the caller.